Add Azure SDK DI doc (#24209)
dotnet/docs: docs/azure/dependency-injection.md

---

> title: Dependency injection with the Azure SDK for .NET

Reviewer: ```suggestion
title: Dependency injection with the Azure .NET SDK
```

description: Learn how to use dependency injection with the Azure SDK for .NET client libraries
ms.date: 05/12/2021

> ms.custom: devx-track-dotnet

Reviewer: ```suggestion
```

Why is this needed? This is in the .NET docs, not the Azure repo - it's my understanding that the `devx-track-*` metadata is only for the Azure docs.

Author: oh I added because other docs seem to have that as well. I'll remove it if it's not needed.

Reviewer 2: Confirmed that it's not needed.

ms.author: pakrym
author: pakrym
---

> # Dependency injection with the Azure SDK for .NET

Reviewer: ```suggestion

# Dependency injection with the Azure .NET SDK
```

> This article demonstrates how to register Azure service clients from the [latest .NET Azure SDKs](https://azure.github.io/azure-sdk/releases/latest/index.html) in an ASP.NET Core application. Every ASP.NET Core application starts by booting up the application using the instructions provided in the `Startup` class. This includes a `ConfigureServices` method that is an ideal place to configure clients.

Reviewer: ```suggestion
This article demonstrates how to register Azure service clients from the [latest Azure .NET SDKs](https://azure.github.io/azure-sdk/releases/latest/index.html) in an ASP.NET Core application. Every ASP.NET Core application starts by booting up the application using the instructions provided in the `Startup` class. This includes a `ConfigureServices` method that is an ideal place to configure clients.
```

{
  services.AddAzureClients(builder =>
  {
    // Add a KeyVault client
    builder.AddSecretClient(keyVaultUrl);

    // Add a Storage account client
    builder.AddBlobServiceClient(storageUrl);

    // Use DefaultAzureCredential by default
    builder.UseCredential(new DefaultAzureCredential());
  });

  services.AddControllers();
}
```

In code above, you need to explicitly specify the `keyVaultUrl` and `storageUrl`. Both variables are `Uri` types. The [Store configuration separately from code](#store-configuration-separately-from-code) section shows how you can avoid specifying the urls explicitly.

The code above uses `DefaultAzureCredential` for authentication. [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/identity/Azure.Identity#defaultazurecredential) will choose the best authentication mechanism based on your environment, allowing you to move your app seamlessly from development to production with no code changes.

## Use the registered clients

With the clients registered in `Startup`, you can now use them:

```csharp
[ApiController]
[Route("[controller]")]
public class MyApiController : ControllerBase
{

>   private readonly BlobServiceClient blobServiceClient;

Reviewer: ```suggestion
  private readonly BlobServiceClient _blobServiceClient;
```

Author: fixed in the latest commit that adds **See also** section

  public MyApiController(BlobServiceClient blobServiceClient)
  {

>     this.blobServiceClient = blobServiceClient;

Reviewer: ```suggestion
    _blobServiceClient = blobServiceClient;
```

  }

  /// Get a list of all the blobs in the demo container
  [HttpGet]
  public async Task<IEnumerable<string>> Get()
  {

>     var containerClient = this.blobServiceClient.GetBlobContainerClient("demo");

Reviewer: ```suggestion
    var containerClient = _blobServiceClient.GetBlobContainerClient("demo");
```

    var results = new List<string>();
    await foreach (BlobItem blob in containerClient.GetBlobsAsync()) 
    {

>       results.Add(blob.Name);

Reviewer: ```suggestion
        results.Add(blob.Name);
```

    }
    return results.ToArray();
  }
}
```

## Store configuration separately from code

> Recall in the [Register client](#register-client) section above, you explicitly specify the `keyVaultUrl` and `storageUrl`. This could cause problems when you run code against different environments during development and production. The ASP.NET Core team suggests [storing such configurations in environment dependent JSON files](/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1#default-configuration). Thus, you can have an `appsettings.Development.json` file with one set of settings and an `appsettings.Production.json` with another set of configurations. The format of the file is:

Reviewer: ```suggestion
In the [Register client](#register-client) section, you explicitly specify the `keyVaultUrl` and `storageUrl`. This could cause problems when you run code against different environments during development and production. The .NET team suggests [storing such configurations in environment dependent JSON files](/dotnet/core/extensions/configuration-providers#json-configuration-provider). Thus, you can have an _appsettings.Development.json_ file with one set of settings and an _appsettings.Production.json_ with another set of configurations. The format of the file is:
```

We should be pointing more broadly to the .NET config docs here. The Azure .NET SDK is not only used in ASP.NET Core apps, it's applicable to all .NET workloads and this could be misleading otherwise.

```json
{
  "AzureDefaults": {
    "Diagnostics": {
      "IsTelemetryDisabled": false,
      "IsLoggingContentEnabled": true
    },
    "Retry": {
      "MaxRetries": 3,
      "Mode": "Exponential"
    }
  },
  "KeyVault": {
    "VaultUri": "https://mykeyvault.vault.azure.net"
  },
  "Storage": {
    "ServiceUri": "https://mydemoaccount.storage.windows.net"
  }
}
```

> You can add any options from the [ClientOptions](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/ClientOptions.cs) into the `AzureDefaults` section. One of the options is the retry policy. See the section [Configure a new try policy](#configure-a-new-retry-policy).

Reviewer: ```suggestion
You can add any options from the <xref:Azure.Core.ClientOptions> into the `AzureDefaults` section. One of the options is the retry policy. For more information, see [Configure a new try policy](#configure-a-new-retry-policy).
```

Use an xref instead, which correctly points to the API ref docs. For related content links, the MS guidance is to phrase it as "For more information, see [Title to the thing](the link)."

Since the `Configuration` object is injected from the host and stored inside the `Startup` constructor, you can do the following:

```csharp
public void ConfigureServices(IServiceCollection services)
{
  services.AddAzureClients(builder =>
  {
    // Add a KeyVault client
    builder.AddSecretClient(Configuration.GetSection("KeyVault"));

    // Add a storage account client
    builder.AddBlobServiceClient(Configuration.GetSection("Storage"));

    // Use DefaultAzureCredential by default
    builder.UseCredential(new DefaultAzureCredential());

    // Set up any default settings
    builder.ConfigureDefaults(Configuration.GetSection("AzureDefaults"));
  });

  services.AddControllers();
}
```

## Configure multiple service clients with different names

Say you have two storage accounts – one for private information and one for public information. Your application transfers data from the public to private storage account after some operation. You need to have two storage service clients. To set this up in `Startup.ConfigureServices`:

```csharp
public void ConfigureServices(IServiceCollection services)
{

>   services.AddAzureClients(builder =>
>   {
>     builder.AddBlobServiceClient(Configuration.GetSection("PublicStorage"));
>     builder.AddBlobServiceClient(Configuration.GetSection("PrivateStorage"))
>       .WithName("PrivateStorage");
>   });

Reviewer: ```suggestion
    services.AddAzureClients(builder =>
    {
        builder.AddBlobServiceClient(Configuration.GetSection("PublicStorage"));
        builder.AddBlobServiceClient(Configuration.GetSection("PrivateStorage"))
            .WithName("PrivateStorage");
    });
```
Four spaces, per style preference for .NET docs.

}
```

> In your controllers, you can access the named service clients using the `IAzureClientFactory`:

Reviewer: ```suggestion
In your controllers, you can access the named service clients using the <xref:Microsoft.Extensions.Azure.IAzureClientFactory%601?displayProperty=nameWithType>:
```

Use xref and qualify type

```csharp
public class HomeControllers : Controller
{

>   private BlobServiceClient publicStorage, privateStorage;

Reviewer: ```suggestion
  private readonly BlobServiceClient _publicStorage;
  private readonly BlobServiceClient _privateStorage;
```

These should be `readonly`, on their own line, and follow the proper naming conventions, leading prefix of `_` for private fields.

  public HomeController(BlobServiceClient defaultClient, IAzureClientFactory<BlobServiceClient> clientFactory)
  {

>     this.publicStorage = defaultClient;
>     this.privateStorage = clientFactory.GetClient("PrivateStorage");

Reviewer: ```suggestion
      _publicStorage = defaultClient;
      _privateStorage = clientFactory.GetClient("PrivateStorage");
```

Author: fixed in the latest commit that adds **See also** section

  }
}
```

The un-named service client is still available in the same way as before. Named clients are additive to this.

## Configure a new retry policy

> At some point, you might want to change the default settings for a service client. You may want different retry settings or to use a different service API version, for example. You can set the retry settings globally or on a per service basis. Say you have the following `appsettings.json` file:

Reviewer: ```suggestion
At some point, you might want to change the default settings for a service client. You may want different retry settings or to use a different service API version, for example. You can set the retry settings globally or on a per service basis. Say you have the following _appsettings.json_ file:
```

Referencing a file name should be italic, instead of code fenced - per MS style guidelines.

```json
{
  "AzureDefaults": {
    "Retry": {
      "maxTries": 3
    }
  },
  "KeyVault": {
    "VaultUri": "https://mykeyvault.vault.azure.net"
  },
  "Storage": {
    "ServiceUri": "https://store1.storage.windows.net"
  },
  "CustomStorage": {
    "ServiceUri": "https://store2.storage.windows.net"
  }
  ```

  You could change the retry policy depending on your needs like so:
  

>   ```csharp
>   public void ConfigureServices(IServiceCollection services)
> {
>   services.AddAzureClients(builder =>
>   {
>     // Establish the global defaults
>     builder.ConfigureDefaults(Configuration.GetSection("AzureDefaults"));
>     builder.UseCredential(new DefaultAzureCredential());
> 
>     // A Key Vault Secrets client using the global defaults
>     builder.AddSecretClient(Configuration.GetSection("KeyVault"));
> 
>     // A Storage client with a custom retry policy
>     builder.AddBlobServiceClient(Configuration.GetSection("Storage"))
>       .ConfigureOptions(options => options.Retry.MaxRetries = 10);
> 
>     // A named storage client with a different custom retry policy
>     builder.AddBlobServiceClient(Configuration.GetSection("CustomStorage"))
>       .WithName("CustomStorage")
>       .ConfigureOptions(options => {
>         options.Retry.Mode = Azure.Core.RetryMode.Exponential;
>         options.Retry.MaxRetries = 5;
>         options.Retry.MaxDelay = TimeSpan.FromSections(120);
>       });
>   });
> }
> ```

Reviewer: ```suggestion
```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddAzureClients(builder =>
    {
        // Establish the global defaults
        builder.ConfigureDefaults(Configuration.GetSection("AzureDefaults"));
        builder.UseCredential(new DefaultAzureCredential());
    
        // A Key Vault Secrets client using the global defaults
        builder.AddSecretClient(Configuration.GetSection("KeyVault"));
    
        // A Storage client with a custom retry policy
        builder.AddBlobServiceClient(Configuration.GetSection("Storage"))
            .ConfigureOptions(options => options.Retry.MaxRetries = 10);
    
        // A named storage client with a different custom retry policy
        builder.AddBlobServiceClient(Configuration.GetSection("CustomStorage"))
            .WithName("CustomStorage")
            .ConfigureOptions(options =>
            {
                options.Retry.Mode = Azure.Core.RetryMode.Exponential;
                options.Retry.MaxRetries = 5;
                options.Retry.MaxDelay = TimeSpan.FromSections(120);
            });
    });
}
```
```

> You can also place policy overrides in the `appsettings.json` file:

Reviewer: ```suggestion
You can also place policy overrides in the _appsettings.json_ file:
```

        builder.ConfigureDefaults(Configuration.GetSection("AzureDefaults"));
        builder.UseCredential(new DefaultAzureCredential());
    
        // A Key Vault Secrets client using the global defaults
        builder.AddSecretClient(Configuration.GetSection("KeyVault"));
    
        // A Storage client with a custom retry policy
        builder.AddBlobServiceClient(Configuration.GetSection("Storage"))
            .ConfigureOptions(options => options.Retry.MaxRetries = 10);
    
        // A named storage client with a different custom retry policy
        builder.AddBlobServiceClient(Configuration.GetSection("CustomStorage"))
            .WithName("CustomStorage")
            .ConfigureOptions(options =>
            {
                options.Retry.Mode = Azure.Core.RetryMode.Exponential;
                options.Retry.MaxRetries = 5;
                options.Retry.MaxDelay = TimeSpan.FromSections(120);
            });
    });
}
```

You can also place policy overrides in the _appsettings.json_ file:

```json
{
  "KeyVault": {
    "VaultUri": "https://mykeyvault.vault.azure.net",
    "Retry": {
      "maxRetries": 10
    }
  }
}
```

## See also

> - [Dependency injection in ASP.NET Core](https://docs.microsoft.com/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.0)
> - [Configuration in ASP.NET Core](https://docs.microsoft.com/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.0)

Reviewer 2: ```suggestion
- [Dependency injection in .NET](/dotnet/core/extensions/dependency-injection)
- [Dependency injection in ASP.NET Core](/aspnet/core/fundamentals/dependency-injection)
- [Configuration in ASP.NET Core](/aspnet/core/fundamentals/configuration)
```

> - [Configuration in ASP.NET Core](https://docs.microsoft.com/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.0)

Author: @scottaddie and @IEvangelist what do you think of linking to these articles?

Reviewer 2: Great idea! See my suggestion below.

Reviewer: I think we should be linking to the .NET specific articles here, both of which do have links to the ASP.NET Core docs if needed. This doc will be used by any workload, not just web apps. As such is should be more broad and encompassing of all .NET workloads where the Azure .NET SDK can be used.

- [Dependency injection in .NET](/dotnet/core/extensions/dependency-injection)
- [Configuration in .NET](/dotnet/core/extensions/configuration)

Author: @IEvangelist sorry for the delay. I think that's a good idea
